x/shield/keeper/proposal.go

```diff
@@ -1,19 +1,616 @@
+	for i := range providers {
+		secureAmt := providers[i].Collateral.ToDec().Mul(lossRatio).TruncateInt()
+		if secureAmt.GT(totalClaimed) {
+			secureAmt = totalClaimed
+		}
+		// Require each provider to secure one more unit, if possible,
+		// so that the last provider does not have to cover combined
+		// truncated amounts.
+		if secureAmt.LT(totalClaimed) && secureAmt.LT(providers[i].Collateral) {
+			secureAmt = secureAmt.Add(sdk.OneInt())
+		}
+		k.SecureFromProvider(ctx, providers[i], secureAmt, duration)
+		totalClaimed = totalClaimed.Sub(secureAmt)
+	}
+
+	// Update purchase states.
+	purchaseList, found := k.GetPurchaseList(ctx, poolID, purchaser)
+	if !found {
+		return types.ErrPurchaseNotFound
+	}
+	var index int
+	for i, entry := range purchaseList.Entries {
+		if entry.PurchaseID == purchaseID {
+			index = i
+			break
+		}
+	}
+	purchase := &purchaseList.Entries[index]
+	if lossAmt.GT(purchase.Shield) {
+		return types.ErrNotEnoughShield
+	}
+	purchase.Shield = purchase.Shield.Sub(lossAmt)
+	votingEndTime := ctx.BlockTime().Add(duration)
+	if purchase.DeletionTime.Before(votingEndTime) {
+		purchase.DeletionTime = votingEndTime
+	}
+	k.SetPurchaseList(ctx, purchaseList)
+
+	// Update pool and global pool states.
+	pool.Shield = pool.Shield.Sub(lossAmt)
+	k.SetPool(ctx, pool)
+
+	totalShield := k.GetTotalShield(ctx)
+	totalShield = totalShield.Sub(lossAmt)
+	k.SetTotalShield(ctx, totalShield)
+	k.SetTotalClaimed(ctx, totalClaimed)
+
 	return nil
 }
 
-func (k Keeper) ClaimUnlock(ctx sdk.Context, proposalID, poolID uint64, loss sdk.Coins) error {
-	return nil
+// SecureFromProvider secures the specified amount of collaterals from
+// the provider for the duration. If necessary, it extends withdrawing
+// collaterals and, if exist, their linked unbondings as well.
+func (k Keeper) SecureFromProvider(ctx sdk.Context, provider types.Provider, amount sdk.Int, duration time.Duration) {
+	// Lenient check:
+	// Check if non-withdrawing, bonded delegation-backed collaterals
+	// cannot cover the amount.
+	if provider.Collateral.Sub(provider.Withdrawing).GTE(amount) && provider.DelegationBonded.GTE(provider.DelegationBonded) {
+		return
+	}
+
+	// Secure the given amount of collaterals until the end of the
+	// lock period by delaying withdrawals, if necessary.
+	endTime := ctx.BlockTime().Add(duration)
+	upcomingWithdrawAmount := k.ComputeWithdrawAmountByTime(ctx, provider.Address, endTime)
+	notWithdrawnSoon := provider.Collateral.Sub(upcomingWithdrawAmount)
+
+	// Secure the given amount of staking (bonded or unbonding) until
+	// the end of the lock period by delaying unbondings, if necessary.
+	totalUnbondingAmount := k.ComputeTotalUnbondingAmount(ctx, provider.Address)
+	upcomingUnbondingAmount := k.ComputeUnbondingAmountByTime(ctx, provider.Address, endTime)
+	notUnbondedSoon := provider.DelegationBonded.Add(totalUnbondingAmount.Sub(upcomingUnbondingAmount))
+
+	if amount.GT(notWithdrawnSoon) || amount.GT(notUnbondedSoon) {
```

Comment:
When will notUnbondedSoon be lesser than notWithdrawnSoon?

Reply:
fixed.